tcpthread.cpp

```diff
@@ -0,0 +1,257 @@
+          if (memcmp(&(connAddr.sin_addr), &(m_remoteAddr.sin_addr), sizeof(struct in_addr)) != 0) {
+            lwarn << "Got a connection from " << connAddrStr
+                  << ", which is not set as a remote." << std::endl;
+            close(m_socket);
+            /* Wait here for some time */
+            std::this_thread::sleep_for(std::chrono::seconds(2));
+            continue;
+          }
+        } else {
+          linfo << "Got a connection from " << connAddrStr << std::endl;
+        }
+        /* At this point we have a valid connection */
+        m_connected = true;
+        /* Clear the old entries in frameBuffer */
+        m_frameBuffer->reset();
+        /* Disable Nagle for this connection */
+        if (setsockopt(m_socket, IPPROTO_TCP, TCP_NODELAY, &nagle, sizeof(nagle))) {
+          lerror << "Could not disable Nagle." << std::endl;
+        }
+      } else {
+        const int nagle = 0;
+        m_socket = socket(AF_INET, SOCK_STREAM, 0);
+        if (m_socket < 0) {
+          lerror << "socket error" << std::endl;
+          continue;
+        }
+        if (setsockopt(m_socket, IPPROTO_TCP, TCP_NODELAY, &nagle, sizeof(nagle))) {
+          lerror << "Could not disable Nagle." << std::endl;
+        }
+        linfo << "Connecting to " << inet_ntoa(m_remoteAddr.sin_addr) << ":" << ntohs(m_remoteAddr.sin_port) << "..." << std::endl;
+        if (connect(m_socket, (struct sockaddr *) &m_remoteAddr, sizeof(m_remoteAddr)) < 0) {
+          close(m_socket);
+          linfo << "Connect failed." << std::endl;
+          /* Wait here for some time */
+          std::this_thread::sleep_for(std::chrono::seconds(2));
+          continue;
+        } else {
+          linfo << "Connected!" << std::endl;
+          m_connected = true;
+        }
+      }
+    } else { /* m_connected == true */
+      /* Prepare readfds */
+      FD_ZERO(&readfds);
+      FD_SET(m_socket, &readfds);
+      FD_SET(m_transmitTimer.getFd(), &readfds);
+      FD_SET(m_blockTimer.getFd(), &readfds);
+      int ret = select(std::max({m_socket, m_transmitTimer.getFd(), m_blockTimer.getFd()})+1,
+        &readfds, NULL, NULL, NULL);
+      if (ret < 0) {
+        if (errno == EOF) {
+          m_connected = false;
+          close(m_socket);
+        }
+        /* Check whether the remote has terminated the connection */
+        lerror << "select error" << std::endl;
+        continue;
+      }
+      if (FD_ISSET(m_transmitTimer.getFd(), &readfds)) {
+        if (m_transmitTimer.read() > 0) {
+          if (m_frameBuffer->getFrameBufferSize())
+            prepareBuffer();
+          else
+            m_transmitTimer.disable();
+        }
+      }
+      if (FD_ISSET(m_blockTimer.getFd(), &readfds)) {
+        m_blockTimer.read();
+      }
+      if (FD_ISSET(m_socket, &readfds)) {
+        /* Clear buffer */
+        memset(buffer, 0, RECEIVE_BUFFER_SIZE);
+        receivedBytes = read(m_socket, buffer, RECEIVE_BUFFER_SIZE);
```

Comment:
This might work on localhost, but on slow connections / changing MTUs, this will cause issues. `read` might return fewer than `RECEIVE_BUFFER_SIZE` and thus `parsePacket` further down will parse something in `buffer` that is not a whole packet.
At this point, it's not possible to know how long the information will be, only that probably smaller or equal than `RECEIVE_BUFFER_SIZE`. With UDP or SCTP in packet mode it is already known by the socket.

The concept of the cannelloni packet is not suited to be ported 1:1 to TCP.
There are at least these options:
* change the read/parse logic here
* create a new protocol for streaming that incorporates the length at the beginning
* transfer CAN frames directly over TCP

# Change parsing logic
The length of the "packet" can only be determined after having read the whole packet, so the part here needs to know whether `parsePacket` is complete.
You call `read` as long as `parsePacket` is not able to parse the packet, i.e. does not throw ```std::runtime_error("Received incomplete packet");```
pseudo-code:
```c
read_total=0
loop {
  n = read(m_socket, buffer+read_total, RECEIVE_BUFFER_SIZE-read_total);
  read_total += n;
  if (read_total > 0) {
    try {
      parseFrames(RECEIVE_BUFFER_SIZE, buffer, ADD_SOME_NOOP, ADD_SOME_NOOP);
    } catch {
       // not done yet
       continue;
    }
    break; // done
}
parsePacket(buffer, receivedBytes, clientAddr);
```
Ideally `parseFrames` should be refactored as catching exceptions to verify that the packet is complete is a bit of an anti-pattern.

# Create a new protocol
That is certainly the best option as it would disconnect the `TCPThread` from the packet concept. Here `length` should be transmitted before any frames and once it is known, the `read` function can loop properly.
A very simple protocol to start from would be e.g. Netstring: https://en.wikipedia.org/wiki/Netstring

# Transfer frames directly
This is a version of the previous option but here can frames would be serialized and directly transferred without encapsulating it into something that resembles a packet.
Again Netstring can be used or a bit simpler: a `read` loop tries to read exactly 5 bytes, then checks the last one (`len`) and continues to read that amount of bytes to parse a whole `canfd_frame`.

Reply:
Thanks for the thorough review. I have realized that the packet concept on top of TCP doesn't make sense and that it's probably the best option to just connect the CAN read to the TCP write calls (almost) directly and let TCP handle the transmission.

Reply:
In any case it would be a good idea that both instances negotiate the protocol during the connect. Something like

client: `CANNELLONI TCPv1\n`
*server checks whether it can speak TCPv1*
server: `OK\n`
both switch to some binary-mode...

Then it would be easier to upgrade at a later point to v2 etc.